### DISCUSSION TITLE
Execute submit on every valid change (onSubmit when valid input after onChange)

### DISCUSSION THREAD

**Discussion by bennettdams** [react-hook-form/react-hook-form#7063] [Q&A] [answered]
Maybe this is easy, but I'm new to RHF.

I have a simple text input that I want to use to filter a list. 
It is a small list, so I want to filter the list on every change instead of e.g. having a dedicated button to submit or on blur.

What is the configuration needed to do this?

I found [this StackOverflow post](https://stackoverflow.com/questions/63466463/how-to-submit-react-form-fields-on-onchange-rather-than-on-submit-using-react-ho), but I'm really unsatisfied with the answers, because I don't want to use `watch`, `trigger` etc. myself, as the chances to make errors are higher.

Maybe I'm thinking in the wrong direction, but instead of handling that myself with `trigger` etc., I would rather want to use `handleSubmit` (and my own `onSubmit` function), because it already does the validation. `onSubmit` should only be executed when the input is valid, so I can rely on that value to filter my list.
I just don't know which combination of `useForm`'s `mode` or `reValidateMode` is working for my use case.

Can anyone help?

[Here's a CodeSandbox](https://codesandbox.io/s/rhf-submit-on-valid-change-tsewt)

...and the code for reference:

```tsx
import React from "react";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import * as z from "zod";

type InputProps = React.ComponentPropsWithoutRef<"input">;

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  function Input(props, ref) {
    return (
      <input
        type="text"
        className="inline-block bg-blue-100 h-12 mt-10 shadow-sm hover:shadow-md rounded-md disabled:bg-gray-100 disabled:cursor-not-allowed"
        {...props}
        ref={ref}
      />
    );
  }
);

function Form(props: React.ComponentPropsWithoutRef<"form">): JSX.Element {
  return (
    <form className="w-full inline-block" {...props}>
      {props.children}
    </form>
  );
}

const schema = z.object({
  username: z.string().min(1, "name required")
});
type Schema = z.infer<typeof schema>;

export default function App() {
  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm<Schema>({
    mode: "all",
    reValidateMode: "onSubmit",
    resolver: zodResolver(schema)
  });

  function onSubmit(data: Schema) {
    console.log("Submitted!", data);
  }

  return (
    <div className="App">
      <h3>React Hook Form example</h3>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <div>
          <Input
            placeholder="Username..."
            id="username"
            autoComplete="off"
            {...register("username")}
          />
          <p>{errors?.username && errors.username.message}</p>
        </div>
      </Form>
    </div>
  );
}
```

**Comment by JamesWong1999** [comment]
```tsx
watch(); // trigger each onChange

if (formState.isValid) {
  onSubmit() // submit with render when form is valid
}
```

**Reply by bennettdams** [reply]
That does not work, because `isValid` is one render cycle behind. So an input that changes a form state from valid to invalid will still be submitted. 

```ts
const data = watch();

const onSubmit = (data: Schema) => console.log("onSubmit data", data);

useEffect(() => {
  console.log("isValid", isValid);
  if (isValid) onSubmit(data);
}, [data, isValid]);
```
With this setup, if you first input a character (like "a") and then remove the character (so the input is an empty string again), will console log the following:

![image](https://user-images.githubusercontent.com/29319414/142188243-ae81bdf0-428d-46ac-806f-eec993e66cf3.png)


As you can see, the empty string is still submitted, even though my Zod schema does not allow an empty string. That's because `isValid` is one render cycle behind.

**Comment by bluebill1049** [comment, marked as answer]
Not sure if this will solve your problem: https://codesandbox.io/s/quiet-pine-chfmz?file=/src/App.tsx

```tsx
import * as React from "react";
import { useForm } from "react-hook-form";
import Headers from "./Header";
import "./styles.css";

type FormValues = {
  firstName: string;
};

let renderCount = 0;

export default function App() {
  const {
    register,
    handleSubmit,
    watch,
    formState,
    formState: { isValidating }
  } = useForm<FormValues>({
    mode: "onChange"
  });
  const onSubmit = (data: FormValues) => console.log(data);
  renderCount++;

  const data = watch();

  React.useEffect(() => {
    if (formState.isValid && !isValidating) {
      console.log(data);
    }
  }, [formState, data, isValidating]);

  return (
    <div>
      <Headers
        renderCount={renderCount}
        description="Performant, flexible and extensible forms with easy-to-use validation."
      />

      <form onSubmit={handleSubmit(onSubmit)}>
        <input
          {...register("firstName", { required: true })}
          placeholder="First Name"
        />
        <input type="submit" />
      </form>
    </div>
  );
}
```

**Reply by bennettdams** [reply]
Thanks, this seems to work, `!isValidating` was the missing piece.

Maybe this is worth writing down as a feature request? It would be nice if e.g. `useForm` would take another option like `autoSubmit`, that executes submit every time the validation was successful. That way you could easily use RHF for a use case I described (like filtering a list). 

Not every input needs a submit button 😅

**Reply by ghost** [reply]
This was a nice example, but could the code sandbox be updated with an action to prove that it submits on change? I gave it a go, and it does not.

**Reply by bennettdams** [reply]
The example shows how to trigger programmatically, but I think instead you can just use `onChange` instead of `onSubmit` on the `<form>` - that way it will always submit when it is valid:

https://codesandbox.io/s/charming-pasteur-kv643z?file=/src/App.tsx

You can see that it works by putting in some text (`console.log`s e.g. "`{firstName: "a"}`") and then remove the text (no `console.log`).

I think this works, not sure if it is recommended though. Same is true for other callbacks on the form like `onBlur`.

**Reply by ghost** [reply]
Forms don't have an `onChange` afaict.

i didn't see the programmatic trigger, and i replied to this post since it was the triggering submission only when validation succeeds that i wanted.  like this part:
```
React.useEffect(() => {
    if (formState.isValid && !isValidating) {
      console.log(data);
    }
  }, [formState, data, isValidating]);
```
except submission actually happens as per your original request
>  It is a small list, so I want to filter the list on every change instead of e.g. having a dedicated button to submit or on blur.

rather than just doing `console.log()`. When i tried it I coudldn't get the right combinations of `useEffect` dependencies to not make it infinitely re-render

**Reply by bennettdams** [reply]
Have you clicked my CodeSandbox? The form works flawlessy with its `onChange` afaik. The benefit of that solution is also that it does not need a `useEffect`, hence no inifinite re-renders and so on.

**Reply by TempleClause** [reply]
@bennettdams 

I tried your CodeSandbox and for input type text it works well. But for some reason it doesn't work for input type checkbox.
See this example:

https://codesandbox.io/s/morning-surf-4c7qrg?file=/src/App.tsx

Any ideas why? 

Thanks!

**Reply by k-zakhariy** [reply]
This is working perfectly for me https://codesandbox.io/s/react-hook-form-auto-save-xgulp?file=/src/AutoSave.tsx

**Reply by musjj** [reply]
With `onChange`, I'm getting the previous value instead of the latest changed one. Anyone else having this issue? I'm using a zod validator, with a few select fields.

**Reply by aleksa-poly-ai** [reply]
@musjj It's related to the micro queue and const/let/var scope. You should use ref to get current (real) value of the variable.

**Reply by CerealeZ** [reply]
It would be great to have the possibility of adding a callback function in the hook options so that at the time of the change, this function is called without the need to perform renderings in the component.

**Comment by craigmcc** [comment]
Is there a reason why you have a form value firstName (as a string) and are trying to render it in a checkbox?  I use boolean form values for that kind of thing.

**Reply by TempleClause** [reply]
Ah that was just a copy paste error ;-) But changing it to a boolean doesn't fix it either 🤷
